Make certain resource deletable even parent is not ready

gemmahou opened the pull request:
### Change description

<!--
Describe what this pull request does.

* If your pull request is to address an open issue, indicate it by specifying the
issue number: 

For example: "Fixes #858"

* For Google internal contributors, you can specify an internal tracking ticket number:

For example: "Fixes b/302708148"

-->
Fixes the deadlock issues on some KCC managed resources caused by the incorrect deletion order
related ticket: b/306583728

### Tests you have done

<!--

Make sure you have run "make ready-pr" to run required tests and ensure this PR is ready to review. 

Also if possible, share a bit more on the tests you have done. 

For example if you have updated the pubsubtopic sample, you can share the test logs from running the test case locally.

go test -v -tags=integration ./config/tests/samples/create -test.run TestAll -run-tests pubsubtopic

-->
Samples test passed on affected resources(AlloydbInstance): https://screenshot.googleplex.com/6ftpm6EhmESEW4s.png
Dynamic test passed: https://screenshot.googleplex.com/5qNmAi5sS5xqKcq.png
`make test` passed.
Reproduced the issue with the fix, the children can be deleted successfully, and then unblock the deletion of their parent.
 
- [x] Run `make ready-pr` to ensure this PR is ready for review.
- [x] Perform necessary E2E testing for changed resources.

gemmahou commented:
Would like to leave some comment about this PR before I'm go on vacation:

In this PR, I added a special handling of the deadlock issue caused by deletion order. The idea is, allow child resources to be deleted, even if the parent resource is not ready.

1. There are various reasons why a parent resource may not be ready (e.g., quota issues, errors from underlying GCP, etc). To prevent breaking changes, I added an allowlist and a logic to limit it to certain resources and certain error(existing children). This fixed known issues with AlloyDB and EdgeContainer.

One downside of this approach is that I need to hard-code the error message string to check for the exact error message. If the service team updates their error message, we will need to update our code.

2. In a prior investigation, we found that we have to resolve the parent in order to delete a child, and we cannot resolve the parent if it is not ready. However, I was able to successfully delete an AlloyDBInstance(child) when the AlloyDBCluster(parent) was in error. I am currently reviewing the code, and since it takes time to reproduce the scenario, I plan to continue investigating whether we need to resolve the parent for deletion when I get back.

3. b/309442012 seems another edge case caused by deletion order. Cx attempted to delete a non-KCC managed resource by removing it from the yaml and re-applying. However, the controller operation here is actually "update" not "delete". I am considering applying a similar allowlist to make child resources updatable even if their parent is not ready. Similarly, I want to limit this to containable resources and check the exact update error message.

justinsb commented on pkg/controller/tf/controller.go:
```diff
@@ -220,15 +220,19 @@ func (r *Reconciler) sync(ctx context.Context, krmResource *krmtotf.Resource) (r
-			orphaned, parent, err := r.isOrphaned(ctx, krmResource)
+		orphaned, parent, err := r.isOrphaned(ctx, krmResource)
+		// Special handling for resources that do not need the orphaned check
+		if !krmtotf.SkipOrphanedCheck(krmResource) {
```
If all resources just deleted via status.selfLink (or equivalent), would this complex logic be needed?

gemmahou replied:
Then we won't need this logic. 
While not all resources have selfLink.  I briefly discussed this approach in my [doc](https://docs.google.com/document/d/1jLQo8clRkHw3gSs9LTMbTTM0iSzgy4fgZunXv2ka_40/edit?pli=1&resourcekey=0-C5aeWmjoceXr4JA9c1GdpQ&tab=t.0#heading=h.cne4eaxl8moi), adding a field called "relative resource name"(similar to selfLink) for all resources and use it for deletion. 
There is some concern over the scope of the fix. Introducing a new field might be "expensive", and most of existing resources do not have the problem. After investigation, only 3-4 resources have this issue on deletion. We may not want to bring a new feature in KCC at this moment. However, I agree that "relative resource name" will not only beneficial for this problem, but for some Cx requests, i.e. https://www.google.com/url?q=https://github.com/GoogleCloudPlatform/k8s-config-connector/issues/295/&sa=D&source=docs&ust=1705093976151315&usg=AOvVaw2SGPnfgQEXAnXOLp2KvkQ9. 
If necessary, I do plan to implement this in the future, and then we can cleanup some logic.

gemmahou commented:
Reply to my previous comment:
`There are various reasons why a parent resource may not be ready (e.g., quota issues, errors from underlying GCP, etc). To prevent breaking changes, I added an allowlist and a logic to limit it to certain resources and certain error(existing children). This fixed known issues with AlloyDB and EdgeContainer.
One downside of this approach is that I need to hard-code the error message string to check for the exact error message. If the service team updates their error message, we will need to update our code.`
Implemented. I will create a ticket to track future actions of this additional logic. i.e. if we need to update a error message, or if we no longer need to handle it and cleanup.

`In a prior investigation, we found that we have to resolve the parent in order to delete a child, and we cannot resolve the parent if it is not ready. However, I was able to successfully delete an AlloyDBInstance(child) when the AlloyDBCluster(parent) was in error. I am currently reviewing the code, and since it takes time to reproduce the scenario, I plan to continue investigating whether we need to resolve the parent for deletion when I get back.`
I believe this [logic](https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/master/pkg/krmtotf/references.go#L193-L195) is the reason.
When we are "deleting" AlloyDBInstance(child), the first state`!deleting` is false, and thus it won't throw the `reference resource not ready` error. It will continue to `resolveTargetFieldValue` and get the resolvedVal of AlloyDBInstance(child) for deletion. I added extra [unit tests](https://github.com/GoogleCloudPlatform/k8s-config-connector/pull/1068/commits/c68a44e73c7afb4173220ea4db5ecabd2a0b0a2a) to cover this logic. 

`b/309442012 seems another edge case caused by deletion order. Cx attempted to delete a non-KCC managed resource by removing it from the yaml and re-applying. However, the controller operation here is actually "update" not "delete". I am considering applying a similar allowlist to make child resources updatable even if their parent is not ready. Similarly, I want to limit this to containable resources and check the exact update error message.`
I do not plan to fix in this PR. The reason is, the Cx is having issue with deleting a non-KCC managed resource. There's nothing much for us to do to handle non-KCC resources. We should suggest them either use "resource reference" to have their resource managed by KCC, or they should make sure they are following the correct order when updating their resources.

jingyih commented on pkg/krmtotf/fetchlivestate.go:
```diff
@@ -72,8 +69,26 @@ func hasEmptySelfLink(resource *Resource) bool {
+	return contains(allowlist, resource.Kind) && k8s.IsDeletionFailureDueToExistingDependent(parent)
+}
+
+func contains(s []string, e string) bool {
```
nit: one option is to use slices.Contains()

gemmahou commented on pkg/controller/tf/controller.go:
```diff
@@ -220,15 +220,19 @@ func (r *Reconciler) sync(ctx context.Context, krmResource *krmtotf.Resource) (r
-		if krmtotf.ShouldResolveParentForDelete(krmResource) {
-			orphaned, parent, err := r.isOrphaned(ctx, krmResource)
+		orphaned, parent, err := r.isOrphaned(ctx, krmResource)
+		// Special handling for resources that do not need the orphaned check
```
Updated the comment

gemmahou commented on pkg/k8s/resource.go:
```diff
@@ -126,6 +126,19 @@ func IsResourceReady(r *Resource) bool {
 	return found && cond.Status == corev1.ConditionTrue
 }
 
+func IsDeletionFailureDueToExistingDependent(r *Resource) bool {
```
I do not think this function is called generally, its only called for allowlisted resources. I agree that we can put this under where it's called.

gemmahou commented on pkg/krmtotf/fetchlivestate.go:
```diff
@@ -53,15 +53,12 @@ func FetchLiveState(ctx context.Context, resource *Resource, provider *tfschema.
-	return !shouldGetImportIDFromSelfForDelete(resource) || hasEmptySelfLink(resource)
+func SkipOrphanedCheck(resource *Resource) bool {
+	allowlist := []string{"KMSCryptoKey"}
+	return contains(allowlist, resource.Kind) || !hasEmptySelfLink(resource)
```
IIUC we have special handling for orphaned resource because we are not able to resolve its parent and then get the unique id for deletion. If the selfLink is non-empty, then we do not need to resolve parent and can simply use selfLink as the unique id and proceed the deletion(see fetchLiveStateFromId). Similar to Justin's comment.

However, that would cause redundant API call, since we assume if a resource is orphaned, then it's no longer exist in the underlying GCP.

maqiuyujoyce replied:
Ah makes sense!

gemmahou commented on pkg/krmtotf/references_test.go:
```diff
@@ -385,6 +386,10 @@ func TestResolveResourceReferenceToTFResource(t *testing.T) {
 					},
 				},
 			},
+			expectedFinalConfig: map[string]interface{}{
```
The error here in this test is "resource not ready"(but exist), not "resource not found". I mention the expectedFinalConfig here to distinct it from the "resource not found" test. It would be preferable for us to verify the specific error message though.

maqiuyujoyce replied:
Gotcha. My concern is that in the test logic, there is no validation of this field, so it can be changed into any value (valid/invalid) without being detected easily.
Are you suggesting that the existence of this field helps readability no matter the value is correct or not? If so, we can leave it as is. If not, then I think we should at least add a TODO to ensure the the value of `expectedFinalConfig` is checked for this test case in the test logic, because accidental changes of this field may harm readability.

gemmahou commented on pkg/krmtotf/references_test.go:
```diff
@@ -642,6 +647,102 @@ func TestResolveResourceReferenceToTFResource(t *testing.T) {
+		referencedResources []*unstructured.Unstructured
+		refConfig           v1alpha1.ReferenceConfig
+		expectedFinalConfig map[string]interface{}
+		shouldError         bool
```
I think I still want to keep `shouldError` here, the goal of the test is to make sure no error under this scenario, `shouldError: true` highlights that. Otherwise I feel this test might be confusing.

maqiuyujoyce replied:
Reading it together with the comment below, it makes sense.

gemmahou commented on pkg/krmtotf/references_test.go:
```diff
@@ -642,6 +647,102 @@ func TestResolveResourceReferenceToTFResource(t *testing.T) {
+			config := tc.config
+			path := strings.Split(tc.refConfig.TFField, ".")
+			if err := ResolveResourceReference(path, config, tc.refConfig, resource, c, smLoader); err != nil {
+				if tc.shouldError {
```
replied in a previous comment

gemmahou commented on pkg/krmtotf/references_test.go:
```diff
@@ -642,6 +647,102 @@ func TestResolveResourceReferenceToTFResource(t *testing.T) {
+			shouldError: false,
+		},
+		{
+			name: "reference not ready",
```
The test scenario is: when we are deleting a resource, and when its reference resource is ready/not ready, there should be no error(source code: https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/master/pkg/krmtotf/references.go#L193). We do not care about if the deletion is successful or not.
The test suite title is "TestResolveResourceReferenceToTFResource_deleting", and the individual test name is "reference is ready" and "reference is not ready".

maqiuyujoyce replied:
Reading it together with the comment above, it makes sense. There is not really a guide for it. I usually prefer to have a single place to describe the test case, i.e. via the `name` field.

gemmahou replied:
Updated the test name to be more descriptive.

gemmahou commented on pkg/krmtotf/references_test.go:
```diff
@@ -642,6 +647,102 @@ func TestResolveResourceReferenceToTFResource(t *testing.T) {
+		shouldError         bool
+	}{
+		{
+			name: "deleting",
```
Replied in another comment

gemmahou commented on pkg/krmtotf/fetchlivestate.go:
```diff
@@ -72,8 +69,26 @@ func hasEmptySelfLink(resource *Resource) bool {
+// when parent exists but has deletion failed error.
+// Due to their API design, the resource is deletable even parent is not ready.
+// See b/306583728#comment8 for details.
+func SkipParentReadyCheckForDeletion(resource *Resource, parent *k8s.Resource) bool {
```
Since we are doing special handling for certain resources. as a helper function, I feel it's straight-forward to return true when the condition meets. I would prefer to negate the condition when we handle it, i.e.
if (!condition) {//do something}
instead of negate it in the helper function.
I also found some similar usage in the code base: [example](https://source.corp.google.com/piper///depot/google3/net/rpc/go/server.go;bpv=1;bpt=1;rcl=599940295;l=3043?q=%22func%20skip%22%20f:.go&sq=package:piper%20file:%2F%2Fdepot%2Fgoogle3%20-file:google3%2Fexperimental). 
I feel this is nothing about correctness but readability, open for suggestions.

gemmahou replied:
Updated the function name and revised the code a bit.

gemmahou commented on pkg/krmtotf/fetchlivestate.go:
```diff
@@ -53,15 +53,12 @@ func FetchLiveState(ctx context.Context, resource *Resource, provider *tfschema.
-
-func ShouldResolveParentForDelete(resource *Resource) bool {
-	return !shouldGetImportIDFromSelfForDelete(resource) || hasEmptySelfLink(resource)
+func SkipOrphanedCheck(resource *Resource) bool {
```
Similar to [this comment](https://github.com/GoogleCloudPlatform/k8s-config-connector/pull/1068/files#r1454250262), I would prefer to negate the condition when we handle it instead of in the helper function.

maqiuyujoyce commented:
@gemmahou I assume you've verified the change works via local integration tests? Didn't see it in the PR description so just want to double check.

gemmahou commented:
> @gemmahou I assume you've verified the change works via local integration tests? Didn't see it in the PR description so just want to double check.

I added screenshots in PR description, ran samples and dynamic tests on affected resource: AlloyDBInstance and it passed. I also reproduced the deletion issues locally. Another resource(EdgeContainerNodePool) in the allowlist is hard to test because it requires physical machine to perform the test and we disabled the samples test on it.

gemmahou commented on pkg/krmtotf/fetchlivestate.go:
```diff
@@ -55,12 +55,13 @@ func FetchLiveState(ctx context.Context, resource *Resource, provider *tfschema.
+func ShouldResolveParentForDelete(resource *Resource) bool {
 	allowlist := []string{"KMSCryptoKey"}
-	return slices.Contains(allowlist, resource.Kind) || !hasEmptySelfLink(resource)
+	return !slices.Contains(allowlist, resource.Kind) || hasEmptySelfLink(resource)
```
To negative the skip function, I simply reverted back to Ziyue's original implementation of ShouldResolveParentForDelete:
```
func shouldGetImportIDFromSelfForDelete(resource *Resource) bool {
	return resource.Kind == "KMSCryptoKey"
}
func ShouldResolveParentForDelete(resource *Resource) bool {
	return !shouldGetImportIDFromSelfForDelete(resource) || hasEmptySelfLink(resource)
}
```
So it returns `!slices.Contains(allowlist, resource.Kind) || hasEmptySelfLink(resource)` 

Seems like my previous skip condition: `return slices.Contains(allowlist, resource.Kind) || !hasEmptySelfLink(resource)` is inaccurate.

maqiuyujoyce replied:
IIUC, the function current means:

If the kind is not KMSCryptoKey, return true;
If the kind is KMSCryptoKey, but it has the empty selfLink field, return true;
If the kind is KMSCryptoKey and it has the non-empty selfLink value, return false.

So it still resolves parent for delete even if it has non-empty selfLink value because it is not a KMSCryptoKey kind.

I think this is different from what the function comment said `We can import the tf state directly from its selfLink instead of sourcing for its parent.`?

gemmahou replied:
I get your point. And thanks for bring it up.
As I said I simply reverted back to the original implementation. the original implementation was only special handling for KMSCryptoKey resource. that's say, if a resource is not KMSCryptoKey, we still want to have complete checks on it, even we could simply delete it via its non-empty selfLink. The comment also mentioned that's special handling of KMSCryptoKey.
While, ideally, if a resource has non-empty selfLink, we could use selfLink as unique id and do not need to resolve its parent. 
What's your thought on this? Shall we keep this function limit to KMSCryptoKey, or expand it to all resource without selfLink?

gemmahou replied:
I personally think limit to KMSCryptoKey would have a minimal impact on our codebase. I am currently discussing with Justin whether we should delete via selfLink or a new "relevantResourceName" field.

maqiuyujoyce replied:
I think to minimize the impact, we should correct the comment to match the condition.

Further improvement can be done in the future.

gemmahou replied:
I revised the comment to be more specific to KMSCryptoKey.

maqiuyujoyce commented:
/lgtm
